bundler/lib/bundler/index.rb

```diff
@@ -46,42 +46,32 @@ def empty?
       true
     end
 
-    def search_all(name)
-      all_matches = specs_by_name(name) # always returns a new Array
-      dupes = @duplicates[name]
-      all_matches.concat(dupes) if dupes
-      @sources.each do |source|
-        all_matches.concat(source.search_all(name))
-      end
-      all_matches
+    def with_all_matching(name, &block)
+      specs_by_name(name)&.each(&block)
+      @duplicates[name]&.each(&block)
+      @sources.each {|source| source.with_all_matching(name, &block) }
     end
 
     # Search this index's specs, and any source indexes that this index knows
     # about, returning all of the results.
     def search(query)
-      results = local_search(query)
-      return results unless @sources.any?
+      results = search_by_query(query) || EMPTY_SEARCH
+      return results if @sources.empty?
 
       @sources.each do |source|
-        results.concat(source.search(query))
+        if results.empty?
+          results = source.search(query)
+        else
+          results.concat(source.search(query))
+        end
       end
       results.uniq(&:full_name)
```

Comment:
Can this be `uniq!` ?

Reply:
Let me see. Tracking the EMPTY_SEARCH as it potentially falls through the nested search determines whether we can modify the result.

Reply:
we could wait to return `EMPTY_SEARCH` until the end, and have `results = results&.concat(...) || ...` (or something like that)

Reply:
I think I figured out a way around it. One moment.